packages/shadycss/src/shadow-parts.js

```diff
@@ -325,11 +331,29 @@ export function onInsertBefore(parentNode, newNode, referenceNode) {
     superRoot === document ? 'document' : superRoot.host.localName;
   const exportMap = findExportedPartMappings(host);
 
+  // Things to keep track of that determine if and how we need to update this
```

Comment:
It's unclear why we need to do runtime work here to support custom properties. Need some explanation of the design. In particular, there's already a runtime system for shimming custom properties, why isn't that sufficient?

Reply:
Hopefully this explains how I'm thinking about it. Starting from what ShadyCSS already does:

1) Template is prepared. Custom properties consumed in this element's `<style>` are discovered.

2) An instance is connected and `styleElement` is called.

3) If the element uses custom properties, their values are computed for this instance. The style cache is checked given those property values, and a new `<style>` is created if there isn't already a suitable one.

Note that at this point in time, we don't know anything about part rules for this instance. We've only dealt with the custom properties that were directly declared in our template styles. It's not immediately obvious how to teach the existing code paths under `styleElement` to handle part styles, because at the point at which it runs, we don't even know what parts the instance has.

4) The instance DOM is first stamped.

5) We check the DOM for parts. We figure out where our host is in the DOM.

6) We check to see if this instance is receiving any part rules that consume a custom property. If so, we have a problem. The values of those properties could be different for each instance, because their values correspond to where the part style is _consumed_, not where it is defined. So our simple-case transformed part rules would have the wrong values.

7) We "adopt" these part rules that have custom properties by adding them to our instance's style info, and we re-run the custom property shim logic that does per-instance styling (the same function as step 3).

Does that help?

One thing that stands out to me from this description is that there is repeated work happening in (3) and (7). There's probably some optimization that could be done where we skip (3) if there's good reason to think that (7) could happen (e.g. because in our ancestry there is at least one part rule that consumes a custom property, even if we don't exactly know where it's going to get applied yet).

WDYT?

Reply:
Yes, that helps. I suggest making sure that (6) is clearly called out in comments in the code.

Reply:
Fleshed out the documentation of `rescopeForCustomProperties` with a better explanation. Also added a TODO and filed https://github.com/webcomponents/polyfills/issues/348 to think about the fact that we are calling `_applyStyleProperties` twice, in effect.